_posts/2019-07-24-pattern-matching-expert-system-classification.md

```diff
@@ -0,0 +1,92 @@
+```
+
+There are some problems with regular expressions that make it a suboptimal choice for describing our rules. They are hard to learn, they quickly become complex and hard to read, and they are too generic and powerful. All of this increases the chance of mistakes.
+
+An interesting alternative is PostgreSQL's [ltree](https://www.postgresql.org/docs/current/ltree.html) extension. It features a language specifically designed to match tree paths. Unfortunately it does not have all the features that we require. And it has been implemented for PostgreSQL only. But if that's what you're using it might be worth taking a look at.
+
+### Tiered Category Expressions
+
+We ended up creating a custom pattern matching language, similar to ltree's, but tailored to our needs. We have named it _Tiered Category Expressions_ (TCEs).
+
+It is designed to be easy to learn for non-technical users, and easy to read and understand. In the simplest scenario an expression mimics the category it is supposed to match, e.g. "fruit > apple" can be matched with the expression `fruit > apple`.
+
+Using a TCE the above rule is expressed as:
+
+```
+>> %fruit% >> apple | apples
+```
+
+TCE features a.o. child and descendant matching, wildcard matching, and negation. Expressions match subtrees by default. Whitespace and special characters are ignored and matching is always case insensitive. Read more about these and other language features in the [TCE language reference](https://github.com/q-m/tiered_category_expressions/blob/master/LANGREF.md). Also TCEs can be transpiled to regular expressions, so the actual matching can be delegated to any platform that supports regular expressions. The Ruby implementation can be found [here](https://github.com/q-m/tiered_category_expressions).
+
+## Conclusions
+
+We only recently started using TCEs with our expert system, though here are some preliminary conclusions and observations.
+
+- It still takes time to write proper rules. This is mostly an up-front task. Once you have a set of good, change-resilient rules, you'll only need to check and update rules when you integrate new sources.
+- Writing rules requires a lot of precision and care. TCE features like wildcards or negation can easily undermine strict unambiguous rules.
+- Non-technical users still need time to get familiar with the TCE syntax.
+- When creating rules, having some validated data provides invaluable insight into real world categories, how they map to target categories, and what exceptions may occur.
+- With a registry of hundreds of rules it becomes hard to detect possible duplicates or overlapping rules (when two rules apply to a single category path).
+- We needed to extend our system to deal with some exceptions. E.g. sometimes a rule must only apply to a specific source and then take precedence over more generic rules.
+- Source data cannot always be trusted to be correct, you might still need some sort of manual validation or error detection in place.
```

Comment:
Is that something worth elaborating on? Maybe guide us through a real-world case where you conclude with validation practices?

Reply:
It's not much more than a suggestion that hasnt really been explored.